.github/workflows/nfr.yml

```diff
@@ -0,0 +1,196 @@
+      test_label:
+        description: 'NFR test to run. Choose between performance, upgrade or all.'
+        required: true
+        default: 'all'
+        type: choice
+        options: [performance, upgrade, all]
+      version:
+        description: 'Version of NGF under test'
+        required: true
+        default: 'edge'
+      nginx-plus:
+        description: 'Run tests with NGINX Plus'
+        required: false
+        default: false
+        type: boolean
+
+defaults:
+  run:
+    shell: bash
+
+concurrency:
+  group: ${{ github.ref_name }}-ci
+  cancel-in-progress: true
+
+permissions:
+  contents: read
+
+jobs:
+  setup-and-run-tests:
+    name: Setup and Run NFR Tests
+    runs-on: ubuntu-22.04
+    permissions:
+      contents: write # needed for opening PR with the results files
+      pull-requests: write # needed for opening PR with the results files
+
+    steps:
+    - name: Checkout Repository
+      uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+
+    - name: Setup Golang Environment
+      uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491 # v5.0.0
+      with:
+        go-version: stable
+
+    - name: Set GOPATH
+      run: echo "GOPATH=$(go env GOPATH)" >> $GITHUB_ENV
+
+    - name: Docker Buildx
+      uses: docker/setup-buildx-action@f95db51fddba0c2d1ec667646a06c2ce06100226 # v3.0.0
+
+    - name: NGF Docker meta
+      id: ngf-meta
+      uses: docker/metadata-action@8e5442c4ef9f78752691e2d8f8d19755c6f78e81 # v5.5.1
+      with:
+        images: |
+          name=gcr.io/${{ secrets.GCP_PROJECT_ID }}/ngf-nfr/nginx-gateway-fabric
+        tags: |
+          type=raw,value=${{ inputs.version }}-${{ github.run_id }}
+
+    - name: NGINX Docker meta
+      id: nginx-meta
+      uses: docker/metadata-action@8e5442c4ef9f78752691e2d8f8d19755c6f78e81 # v5.5.1
+      with:
+        images: |
+          name=gcr.io/${{ secrets.GCP_PROJECT_ID }}/ngf-nfr/nginx-gateway-fabric/${{ inputs.nginx-plus == true && 'nginxplus' || 'nginx' }}
+        tags: |
+          type=raw,value=${{ inputs.version }}-${{ github.run_id }}
+
+    - name: Build binary
+      uses: goreleaser/goreleaser-action@7ec5c2b0c6cdda6e8bbb49444bc797dd33d74dd8 # v5.0.0
+      with:
+        version: latest
+        args: build --snapshot --clean
+
+    - name: Authenticate to Google Cloud
+      id: auth
+      uses: google-github-actions/auth@a6e2e39c0a0331da29f7fd2c2a20a427e8d3ad1f # v2.1.1
+      with:
+        token_format: access_token
+        workload_identity_provider: ${{ secrets.GCP_WORKLOAD_IDENTITY }}
+        service_account: ${{ secrets.GCP_SERVICE_ACCOUNT }}
+
+    - name: Set up Cloud SDK
+      uses: google-github-actions/setup-gcloud@98ddc00a17442e89a24bbf282954a3b65ce6d200 # v2.1.0
+      with:
+        project_id: '${{ secrets.GCP_PROJECT_ID }}'
+        install_components: 'kubectl'
+
+    - name: Login to GCR
+      run: gcloud auth configure-docker gcr.io -q
+
+    - name: Build NGF Docker Image
```

Comment:
Would be nice if we could have these common steps somewhere else and be imported in some way. We build binary/images in three different places now I think.

Reply:
We have a main build.yml that we use for the ci.yml but it's got an awful lot of additional steps we don't require here that I took the simpler route 😅 I'll take a look at deduping though, it would be a good idea.

Reply:
Do we need to build the images for these tests? Can we pull them from ghcr instead?

Reply:
Hmmm, good point. I could probably skip the build steps unless the NGINX image is plus.

Reply:
Likely outside of the scope of this PR now, but we have images being built in the regular build step, when we run helm tests, when we run conformance tests, and now @salonichf5 is adding it to the functional test pipeline.

Having the single build job that other jobs/workflows could wait for and then use the images from would be nice, though I'm not sure how difficult that is in certain contexts. I know I tried it a little bit with a third-party workflow I found to wait for some release steps to happen before running conformance tests, but it had some issues.